semver/validity.ts

```diff
@@ -0,0 +1,162 @@
+ * @returns True if value is a valid SemVer otherwise false
+ */
+export function isSemVer(value: unknown): value is SemVer {
+  if (value == null) return false;
+  if (Array.isArray(value)) return false;
+  if (typeof value !== "object") return false;
+  if (value === INVALID) return true;
+  if (value === ANY) return true;
+
+  const { major, minor, patch, build, prerelease } = value as Record<
+    string,
+    unknown
+  >;
+  const result = typeof major === "number" && isValidNumber(major) &&
+    typeof minor === "number" && isValidNumber(minor) &&
+    typeof patch === "number" && isValidNumber(patch) &&
+    Array.isArray(prerelease) &&
+    Array.isArray(build) &&
+    prerelease.every((v) => typeof v === "string" || typeof v === "number") &&
+    prerelease.filter((v) => typeof v === "string").every((v) =>
+      isValidString(v)
+    ) &&
+    prerelease.filter((v) => typeof v === "number").every((v) =>
+      isValidNumber(v)
+    ) &&
+    build.every((v) => typeof v === "string" && isValidString(v));
+  return result;
+}
+
+/**
+ * Does a deep check on the value to see if it is a valid SemVerComparator object.
+ *
+ * Objects with extra fields are still considered valid if they have at
+ * least the correct fields.
+ *
+ * Adds a type assertion if true.
+ * @param value The value to check if its a SemVerComparator
+ * @returns True if the object is a SemVerComparator otherwise false
+ */
+export function isSemVerComparator(value: unknown): value is SemVerComparator {
+  if (value == null) return false;
+  if (value === NONE) return true;
+  if (value === ALL) return true;
+  if (Array.isArray(value)) return false;
+  if (typeof value !== "object") return false;
+  const { operator, semver, min, max } = value as SemVerComparator;
+  return (
+    isValidOperator(operator) &&
+    isSemVer(semver) &&
+    isSemVer(min) &&
+    isSemVer(max)
+  );
+}
+
+/**
+ * Does a deep check on the object to determine if its a valid range.
+ *
+ * Objects with extra fields are still considered valid if they have at
+ * least the correct fields.
+ *
+ * Adds a type assertion if true.
+ * @param value The value to check if its a valid SemVerRange
+ * @returns True if its a valid SemVerRange otherwise false.
+ */
+export function isSemVerRange(value: unknown): value is SemVerRange {
+  if (value == null) return false;
+  if (Array.isArray(value)) return false;
+  if (typeof value !== "object") return false;
+  const { ranges } = value as SemVerRange;
+  return (
+    Array.isArray(ranges),
+      ranges.every((r) =>
+        Array.isArray(r) && r.every((c) => isSemVerComparator(c))
+      )
+  );
+}
+
+/**
+ * Checks to see if the value is a valid Operator string.
+ *
+ * Adds a type assertion if true.
+ * @param value The value to check
+ * @returns True if the value is a valid Operator string otherwise false.
+ */
+export function isValidOperator(value: unknown): value is Operator {
```

Comment:
I think this and the below 2 functions have little value to the users. How about making them private/unexported?